### PR TITLE
Fix vstest.console.exe grabs exclusive read access to its test container

### PR DESCRIPTION
**Pull request by smadala**
## Description
- Use FileStream API to create stream.
- Default FileShare for [FileStream .ctor](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/IO/FileStream.cs#L15) is FileShare.Read where as [File.Open()](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/File.cs#L146) default FileShare value is FileShare.None

## Related issue
https://github.com/Microsoft/vstest/issues/1638

### CODE REVIEW

**Review by abhishkk — APPROVED** [review]

File: src/Microsoft.TestPlatform.ObjectModel/Navigation/DiaSession.cs [hunk 1]
```diff
@@ -99,7 +99,7 @@ private static ISymbolReader GetSymbolReader(string binaryPath)
             // For remote scenario, also look for pdb in current directory, (esp for UWP)
             // The alternate search path should be an input from Adapters, but since it is not so currently adding a HACK
             pdbFilePath = !File.Exists(pdbFilePath) ? Path.Combine(Directory.GetCurrentDirectory(), Path.GetFileName(pdbFilePath)) : pdbFilePath;
-            using (var stream = new FileHelper().GetStream(pdbFilePath, FileMode.Open, FileAccess.Read))
+            using (var stream = new FileHelper().GetStream(pdbFilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
```
Comment (cltshivash) [inline src/Microsoft.TestPlatform.ObjectModel/Navigation/DiaSession.cs]: Possibility to have a single method for GetStreamForRead in FileHelper

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs [hunk 2]
```diff
@@ -11,30 +11,39 @@ namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors.Utilities
...
     internal class AssemblyMetadataProvider : IAssemblyMetadataProvider
     {
         private static AssemblyMetadataProvider instance;
+        private IFileHelper fileHelper;
 
         /// <summary>
         /// Gets the instance.
         /// </summary>
-        internal static AssemblyMetadataProvider Instance => instance ?? (instance = new AssemblyMetadataProvider());
+        internal static AssemblyMetadataProvider Instance => instance ?? (instance = new AssemblyMetadataProvider(new FileHelper()));
```
Comment (cltshivash) [inline src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: why internal.. make it public.. the class is internal already

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs [hunk 3]
```diff
@@ -11,30 +11,39 @@ namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors.Utilities
...
+
+        internal AssemblyMetadataProvider(IFileHelper fileHelper)
```
Comment (cltshivash) [inline src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: same comment..

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs [hunk 4]
```diff
@@ -11,30 +11,39 @@ namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors.Utilities
...
+                using (var assemblyStream = this.fileHelper.GetStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
                 {
-                    frameworkName = GetFrameworkNameFromAssemblyMetadata(assemblyStream);
+                    frameworkName = AssemblyMetadataProvider.GetFrameworkNameFromAssemblyMetadata(assemblyStream);
                 }
             }
             catch (Exception ex)
             {
-                EqtTrace.Warning("GetFrameWorkFromMetadata: failed to determine TargetFrameworkVersion: {0} for assembly: {1}", ex, filePath);
+                EqtTrace.Warning("AssemblyMetadataProvider.GetFrameWork: failed to determine TargetFrameworkVersion: {0} for assembly: {1}", ex, filePath);
```
[the remark this answers is not in the document]
Comment (cltshivash) [reply src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: The parameters are interchanged.. also please inline it.

[the remark this answers is not in the document]
Comment (smadala) [reply src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: >TargetFrameworkVersion: {0} Here {0} is ex and not TargetFrameworkVersion. Can you have a look?

Done.

>The parameters are interchanged.. also please inline it.

Making inline causes additional statement string.format which we don't want be cause most of time we don't in diag mode.  See the below IL code.
image

File: test/vstest.console.PlatformTests/AssemblyMetadataProviderTests.cs [hunk 5]
```diff
@@ -135,6 +161,7 @@ public void GetFrameWorkForNativeDll()
 
         private void TestDotnetAssemblyArch(string projectName, string framework, Architecture expectedArch, long expectedElapsedTime)
         {
+            this.isManagedAssemblyArchitectureTest = true;
```
[the remark this answers is not in the document]
Comment (smadala) [reply test/vstest.console.PlatformTests/AssemblyMetadataProviderTests.cs]: Actually now we don't need FileShare. Added `FileHelperTests.cs` to add required test.

File: src/Microsoft.TestPlatform.ObjectModel/Navigation/DiaSession.cs (hunk 1 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.ObjectModel/Navigation/DiaSession.cs]: This change not required now. we don't want to add new method which leads to additional learning . The helper is just wrapper to IO apis.

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs (hunk 2 above)
Comment (smadala) [reply src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: Done.

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs (hunk 3 above)
Comment (smadala) [reply src/vstest.console/CommandLine/AssemblyMetadataProvider.cs]: This we don't expect to used by any other assembly except test assembly. If we make this public we will lose the flexibility to change the signature.

**Pull request merged**